cmd/agent/main.go

```diff
@@ -5,16 +5,44 @@
 
 //go:build !windows
 
+// Main package for the agent binary
 package main
 
 import (
+	"fmt"
 	"os"
+	"path"
+	"strings"
 
 	"github.com/DataDog/datadog-agent/cmd/agent/command"
 	"github.com/DataDog/datadog-agent/cmd/agent/subcommands"
 	"github.com/DataDog/datadog-agent/cmd/internal/runcmd"
+	"github.com/spf13/cobra"
 )
 
+func init() {
+	registerAgent(func() *cobra.Command {
+		return command.MakeCommand(subcommands.AgentSubcommands())
+	}, "agent", "datadog-agent")
+}
+
+var agents = map[string]func() *cobra.Command{}
+
+func registerAgent(getCommand func() *cobra.Command, names ...string) {
+	for _, name := range names {
+		agents[name] = getCommand
+	}
+}
+
 func main() {
-	os.Exit(runcmd.Run(command.MakeCommand(subcommands.AgentSubcommands())))
+	executable := path.Base(os.Args[0])
```

Comment:
Passing the program name as the first parameter is a common convention but this is not a requirement: https://stackoverflow.com/questions/2050961/is-argv0-name-of-executable-an-accepted-standard-or-just-a-common-conventi.
```c
#include <unistd.h>
int main(int argc, char* argv[])
{
    execl("./bin/agent/agent", "SOMETHING", 0);
    return 0;
}
```
Running the previous code set `os.Args[0]` to `SOMETHING`. 

Also, there is an issue if the user creates its own alias like `dd-process-agent` for the `process-agent`.  

I am pretty sure some users will have an issue but it is hard to evaluate the number. One way to do that is to send a telemetry for `os.Args[0]` and count how much time it doesn't match the program name.

I think this is OK for hosts where there is no previous Agents installed as in case of an issue, the user will investigate. If single agent binary can replace existing working Agents then it might be an issue.

Reply:
I agree that using on first element of argv with an hardcoded array is quite unreliable...

Can't we have the single binary (named eg. `main`) taking as first argument the name of the agent to run, and then one small script for each agent which just calls `main` with the right first arg ?
Ie. a `security-agent` script which would just `exec ./main security-agent "$@"`, etc

Might need to add a symbolic link to make sure it doesn't break if `security-agent` is moved out of the directory...

Reply:
> one small script for each agent 

@pgimalac 
If you think about something like a bash script, this will not work well. For example in python, running a script requires `shell=True` when using  `subprocess.run`.

Another option is to have one small binary per agent that plays the same role as your script for example using `exec.Command("main", "security-agent")`. The drawback of that is you have two running processes per agent instead of one. It is unclear if it can have any side effect.
One idea to avoid having two processes is to use [go plugin](https://pkg.go.dev/plugin). There is one minimal binary per agent, that loads the plugin. The number of running processes is the same (no two running processes per agent). Unfortunately, it would significantly increase the complexity of the build.

Reply:
For the first part I'm not sure I understood as I just tested it and it works.
For the second part we can still use the `syscall.Exec` to replace the actual process.
Didn't know about go plugin I'll have to look into that !

Reply:
EDIT: for the first part we discussed it offline and the misunderstanding was that the script needs a `shebang` at the start so that the OS knows how to execute it

Reply:
Indeed, in my test I didn't use `shebang` and it works when using it. I am just wondering it this is something that works on all our supported distributions.

Reply:
Using `exec ./main security-agent "$@"` will have an impact on the way the agent parses the command line as the first argument is `security-agent`.
You can use `-a` to specify the zeroth argument to COMMAND.
```
exec -a security-agent ./main  "$@"
```
`exec` replaces the current shell meaning that if you run the agent from the shell it will be destroyed when the Agent terminates which is something we want.

Reply:
@ogaca-dd  Thanks. Unfortunately `exec -a` is a bashism and bash may not be present so we cannot use it

Reply:
The Go plugin mechanism is sadly full of caveats, one significant one being that the portability is not super good (https://pkg.go.dev/plugin#hdr-Warnings)

Reply:
As discussed privately, using a env variable is probably the easiest solution with the less drawbacks.